pulsar-broker/src/main/java/org/apache/pulsar/broker/service/HashRangeAutoSplitStickyKeyConsumerSelector.java

```diff
@@ -18,86 +18,64 @@
-    private final Map<Consumer, Integer> consumerRange;
+    // Consistent-Hash ring
+    private final NavigableMap<Integer, Consumer> hashRing;
 
-    public HashRangeAutoSplitStickyKeyConsumerSelector() {
-        this(DEFAULT_RANGE_SIZE);
-    }
+    private final int numberOfPOints;
 
-    public HashRangeAutoSplitStickyKeyConsumerSelector(int rangeSize) {
-        if (rangeSize < 2) {
-            throw new IllegalArgumentException("range size must greater than 2");
-        }
-        if (!is2Power(rangeSize)) {
-            throw new IllegalArgumentException("range size must be nth power with 2");
-        }
-        this.rangeMap = new ConcurrentSkipListMap<>();
-        this.consumerRange = new HashMap<>();
-        this.rangeSize = rangeSize;
+    public HashRangeAutoSplitStickyKeyConsumerSelector(int numberOfPOints) {
+        this.hashRing = new TreeMap<>();
+        this.numberOfPOints = numberOfPOints;
     }
 
     @Override
-    public synchronized void addConsumer(Consumer consumer) throws ConsumerAssignException {
-        if (rangeMap.size() == 0) {
-            rangeMap.put(rangeSize, consumer);
-            consumerRange.put(consumer, rangeSize);
-        } else {
-            splitRange(findBiggestRange(), consumer);
+    public void addConsumer(Consumer consumer) throws ConsumerAssignException {
+        rwLock.writeLock().lock();
+        try {
+            // Insert multiple points on the hash ring for every consumer
+            // The points are deterministically added based on the hash of the consumer name
+            for (int i = 0; i < numberOfPOints; i++) {
+                String key = consumer.consumerName() + i;
+                int hash = Murmur3_32Hash.getInstance().makeHash(key.getBytes());
+                hashRing.put(hash, consumer);
+            }
+        } finally {
+            rwLock.writeLock().unlock();
         }
     }
 
     @Override
-    public synchronized void removeConsumer(Consumer consumer) {
-        Integer removeRange = consumerRange.remove(consumer);
-        if (removeRange != null) {
-            if (removeRange == rangeSize && rangeMap.size() > 1) {
-                Map.Entry<Integer, Consumer> lowerEntry = rangeMap.lowerEntry(removeRange);
-                rangeMap.put(removeRange, lowerEntry.getValue());
-                rangeMap.remove(lowerEntry.getKey());
-                consumerRange.put(lowerEntry.getValue(), removeRange);
-            } else {
-                rangeMap.remove(removeRange);
+    public void removeConsumer(Consumer consumer) {
+        rwLock.writeLock().lock();
+        try {
+            // Remove all the points that were added for this consumer
+            for (int i = 0; i < numberOfPOints; i++) {
+                String key = consumer.consumerName() + i;
+                int hash = Murmur3_32Hash.getInstance().makeHash(key.getBytes());
+                hashRing.remove(hash, consumer);
```

Comment:
When removing consumers from the hash ring, is it necessary to check if the value of the key we are removing is also equals the consumer?  In what case, will that not be true?

Reply:
It's necessary because there can be a hash conflict. When that happens, the point will only be assigned to one consumer. 

It is not a problem since we have multiple points in the ring, though when we remove `c1` we need to be sure we're not removing a point that was actually assigned to `c2`.

Reply:
I see thanks for the explanation